### PR TITLE
Extend API Client to allow retries on failed requests

### PR DESCRIPTION
**Pull request by james-harper**
### Added
- Added `RetryApiClient` which extends `ApiClient` to retry failed requests multiple times

https://www.pivotaltracker.com/story/show/143412155

### CODE REVIEW

**Comment by james-harper** [comment]
@wnowicki I have made the changes based on our discussion. There is now an AbstractRetryApiClient and an implementation of that

File: src/AbstractApiClient.php [hunk 1]
```diff
@@ -40,12 +40,24 @@
...
     /**
+     * Initialise a Guzzle Client which uses the retry middleware
+     *
+     * @author JH
+     * @param array $config
+     * @return Client
+     */
+    protected function initialiseClient(array $config = [])
+    {
+        return new Client($config);
```
Comment (wnowicki) [inline src/AbstractApiClient.php]: is there any contract (interface) on this client?

File: src/AbstractRetryApiClient.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Initialise a Guzzle Client which uses the retry middleware
+     *
+     * @author JH
+     * @param array $config
+     * @return Client
+     */
+    protected function initialiseClient(array $config = [])
+    {
+        if (!isset($config['handler'])) {
```
Comment (wnowicki) [inline src/AbstractRetryApiClient.php]:
```php
!isset($config['handler']) || !$config['handler'] instanceof HandlerStack
```

??

File: src/AbstractRetryApiClient.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    abstract protected function retryDecider();
+
+    /**
+     * The number of times to retry an unsuccessful request
+     *
+     * @author JH
+     * @return int
+     */
+    abstract protected function getMaxRetries();
```
Comment (wnowicki) [inline src/AbstractRetryApiClient.php]: maybe this one with default implementation ?

**Comment by james-harper** [comment]
@wnowicki I have made some changes based on your comments

**Pull request merged**